src/LiveComponent/src/Normalizer/LiveComponentNormalizer.php

```diff
@@ -0,0 +1,257 @@
+    {
+        // todo - lazy
+    }
+
+    /**
+     * @param object $component
+     */
+    public function normalize(mixed $component, string $format = null, array $context = []): array
+    {
+        foreach (AsLiveComponent::preDehydrateMethods($component) as $method) {
+            $component->{$method->name}();
+        }
+
+        $data = [];
+        $readonlyProperties = [];
+        $frontendPropertyNames = [];
+
+        foreach (AsLiveComponent::liveProps($component) as $propContext) {
+            $property = $propContext->reflectionProperty();
+            $liveProp = $propContext->liveProp();
+            $name = $property->getName();
+            $frontendName = $liveProp->calculateFieldName($component, $property->getName());
+
+            if (isset($frontendPropertyNames[$frontendName])) {
+                $message = sprintf('The field name "%s" cannot be used by multiple LiveProp properties in a component. Currently, both "%s" and "%s" are trying to use it in "%s".', $frontendName, $frontendPropertyNames[$frontendName], $name, \get_class($component));
+
+                if ($frontendName === $frontendPropertyNames[$frontendName] || $frontendName === $name) {
+                    $message .= sprintf(' Try adding LiveProp(fieldName="somethingElse") for the "%s" property to avoid this.', $frontendName);
+                }
+
+                throw new \LogicException($message);
+            }
+            $frontendPropertyNames[$frontendName] = $name;
+
+            if ($liveProp->isReadonly()) {
+                $readonlyProperties[] = $frontendName;
+            }
+
+            // TODO: improve error message if not readable
+            $value = $this->propertyAccessor->getValue($component, $name);
+            $dehydratedValue = null;
+
+            if ($method = $liveProp->dehydrateMethod()) {
+                // TODO: Error checking
+                $dehydratedValue = $component->$method($value);
+            } elseif (\is_object($dehydratedValue = $value)) {
+                $dehydratedValue = $this->normalizer->normalize($dehydratedValue, $format, [self::CONTEXT_KEY => self::CONTEXT_NESTED]);
+            }
+
+            if (\count($liveProp->exposed()) > 0) {
+                $data[$frontendName] = [
+                    self::EXPOSED_PROP_KEY => $dehydratedValue,
+                ];
+
+                foreach ($liveProp->exposed() as $propertyPath) {
+                    $value = $this->propertyAccessor->getValue($component, sprintf('%s.%s', $name, $propertyPath));
+                    $data[$frontendName][$propertyPath] = \is_object($value) ? $this->normalizer->normalize($value, $format, [self::CONTEXT_KEY => self::CONTEXT_NESTED]) : $value;
+                }
+            } else {
+                $data[$frontendName] = $dehydratedValue;
+            }
+        }
+
+        $data[self::CHECKSUM_KEY] = $this->computeChecksum($data, $readonlyProperties);
+
+        return $data;
+    }
```

Comment:
What's the "profit" of moving this logic here vs keeping it inside of `LiveComponentHydrator`. Right now, it looks like we are simply setting a special context flag... then calling the normalizer so that the normalizer can call this class exactly. I like that the individual properties of the component go through the (de)normalizer. But I'm not sure about moving this main logic here. WDYT?

Reply:
My thinking was to eventually remove `LiveComponentHydrator`. Keeping it for now helped with the refactor. When I went to remove it, I saw how ugly it was to use the symfony serializer directly (and we need to do this in multiple places) so kept it for now.

I want to see how #264 pans out before I continue this.

Reply:
Ah, you're thinking to call the serializer directly form `LiveComponentSubscriber`? Grab the raw JSON string from there and pass it straight to the serializer system? That is, indeed, interesting

Reply:
Yes, same with `LiveComponentRuntime`. I want to see how that looks.

Reply:
With attributes/mounted component I discovered I couldn't easily have the `LiveComponentNormalizer`, I've kept the logic in `LiveComponentHydrator`.